src/Stream.ts

```diff
@@ -0,0 +1,108 @@
+import { Effect } from "./Effect";
+import * as Fx from "./Fx";
+import * as Task from "./Task";
+import * as Async from "./Async";
+import * as Fail from "./Fail";
+import * as Fork from "./Fork";
+
+export class Stream<A> extends Effect('Stream')<A, void> { }
+
+export namespace Stream { 
+  export type Event<T> = T extends Stream<infer A> ? A : never
+}
+
+export function event<const A>(a: A): Fx.Fx<Stream<A>, void> { 
+  return new Stream(a)
+}
+
+export function observe<E, R, E2>(fx: Fx.Fx<E, R>, f: (a: Stream.Event<E>) => Fx.Fx<E2, void>): Fx.Fx<Exclude<E, Stream<any>> | E2, R> { 
+  return fx.pipe(Fx.handle<typeof Stream, E2>(Stream, (a) => f(a as Stream.Event<E>)))
+} 
+
+export function collectAll<E, A>(fx: Fx.Fx<E, A>): Fx.Fx<Exclude<E, Stream<any>>, readonly Stream.Event<E>[]> {
+  return Fx.fx(function* () {
+    const events: Stream.Event<E>[] = []
+    yield* observe(fx, a => { 
+      events.push(a)
+      return Fx.unit
+    })
+    return events
+  })
+}
+
+export function map<E, A, B>(fx: Fx.Fx<E, A>, f: (a: Stream.Event<E>) => B): Fx.Fx<Exclude<E, Stream<any>> | Stream<B>, A> { 
```

Comment:
I'm not sure it's really possible to accurately separate the events without creating different Effects though right?

Reply:
In my head, it's possible 😂 .  I think it requires 2 functions, though, a type guard and a mapping function.  The same has proven to be true with open variants at my day job.  Since effects are basically open variants, I think it would turn out to be similar.

That said, I think going with how you've done it here is the right thing for a foundational `map`.

Reply:
Ahh yes, chaining filter and map should work now that it has the refinement overload to get a similar behavior.

If we wanted to introduce an Option/Maybe type and I can understand why we wouldn't, but FWIW I really like utilizing the `A => Option<B>` signature to sorta merge those two concepts. I think any ADTs naturally have functions which take the ADT and then zooms in on a particular type parameter or state of the ADT that can be modeled `ADT<...> => Option<B>` in one go instead of needing `filter(ADT.isSomeStateOrValue).map(ADT.getSomeStateOrValue)` to just `filterMap(ADT.getSomeStateOrValue)`

Reply:
At least in the case of open variants, it's a little different.  For example, given `Fx<Stream<A> | Stream<B>, ...>`, filtering with `(x: A | B) => x is A` _discards_ the Bs.  In contrast, an operation like `map(p: (x: A | B) => x is A, f: (a: A) => C, ...)` can map the As to Cs while also retaining the Bs.  It's similar to how handlers work, actually, by handling some effects and allow others to propagate.